README.md
## ProGenSys — problem set generation system

This is a micro-LMS (learning management system) has been used over the years to automatically generate problem sets for the “Discrete Structures” course, grade the works and allow students to track the progress. A course is associated with a default trajectory (a collection of topics) that the student should go through. The students get problems sets, each problem set is unique for each student. An automated system employs combinatorial optimization to create problem sets for the students so that the students get different (where possible) problems while receiving problems from the right topics that they have yet to cover.

misc_files/history.txt
14 10 3 1
14 13 3 1
7 3 3 1
31 6 3 1
42 6 3 1
17 7 1 1
7 12 2 1
23 1 2 1
19 4 2 1
40 3 3 1
20 7 3 1
22 6 3 1
21 6 3 1
8 5 3 1
23 7 2 1
11 11 1 1
33 3 3 1
11 5 3 1
38 1 3 1
12 6 3 1
15 11 1 1
14 1 3 1
13 7 3 1
27 3 1 1
15 4 3 1
39 11 1 1
5 3 2 1
26 12 3 1
3 2 3 1
2 6 3 1
5 11 1 1
4 5 1 1
27 5 1 1
38 3 3 1
18 4 3 1
7 5 3 1
19 13 1 1
24 4 2 1
32 3 3 1
31 7 3 1
8 12 1 1
22 12 2 1
33 4 3 1
32 12 1 1
36 5 3 1
34 3 2 1
36 2 3 1
38 5 3 1
37 3 3 1
15 13 3 1
13 13 2 1
31 11 3 1
25 5 1 1
7 11 3 1
3 11 1 1
2 1 3 1
26 2 3 1
4 12 1 1
28 1 3 1
2 12 3 1
5 1 3 1
29 4 1 1
27 2 1 1
16 7 3 1
31 1 3 1
1 11 3 1
42 11 1 1
41 1 2 1
19 7 2 1
18 5 2 1
17 11 1 1
24 3 3 1
32 2 3 1
40 1 3 1
22 7 2 1
21 5 3 1
8 6 3 1
23 6 3 1
37 12 1 1
9 7 1 1
33 2 3 1
23 2 3 1
11 4 3 1
35 3 3 1
12 7 3 1
36 4 3 1
43 2 3 1
42 12 1 1
13 6 3 1
15 7 3 1
6 7 3 1
26 13 1 1
14 6 2 1
4 11 3 1
3 5 1 1
2 7 3 1
16 13 1 1
4 6 1 1
28 11 1 1
6 1 2 1
30 2 3 1
27 6 1 1
2 4 3 1
7 4 3 1
43 1 2 1
19 12 1 1
42 1 2 1
6 13 3 1
24 7 1 1
19 1 3 1
18 11 1 1
18 7 0 1
18 9 3 1
16 3 3 1
24 2 3 1
32 1 3 1
8 13 1 1
22 13 2 1
9 8 3 1
32 13 3 1
37 13 1 1
28 4 2 1
36 3 3 1
37 2 3 1
15 12 2 1
39 3 3 1
13 12 2 1
25 3 3 1
25 4 1 1
40 12 1 1
26 3 3 1
43 4 3 1
41 12 1 1
4 13 1 1
28 2 3 1
38 2 3 1
30 5 3 1
29 3 1 1
40 5 3 1
16 9 3 1
17 5 3 1
16 11 1 1
19 6 1 1
18 2 2 1
42 7 1 1
42 9 3 1
7 13 1 1
6 11 1 1
7 7 3 1
22 4 3 1
21 4 3 1
8 7 1 1
32 4 3 1
22 11 2 1
9 6 3 1
33 1 3 1
11 7 3 1
35 2 2 1
34 6 2 1
37 11 1 1
11 8 3 1
14 7 2 1
13 5 2 1
1 1 3 1
15 6 3 1
39 5 3 1
1 7 3 1
25 2 2 1
3 4 2 1
40 11 1 1
4 7 1 1
5 7 3 1
6 6 3 1
5 6 3 1
43 3 3 1
16 2 3 1
17 3 3 1
41 2 3 1
40 2 3 1
16 1 2 1
30 3 3 1
22 2 3 1
8 1 3 1
23 11 3 1
35 11 1 1
34 1 2 1
11 1 3 1
24 1 3 1
37 1 2 1
39 2 3 1
13 11 1 1
25 11 2 1
1 13 2 1
28 3 3 1
29 2 3 1
6 12 2 1
29 11 1 1
17 4 2 1
16 12 1 1
18 3 3 1
42 4 2 1
43 5 3 1
20 2 3 1
22 5 3 1
21 3 3 1
41 11 3 1
32 5 3 1
22 8 2 1
9 5 3 1
23 13 1 1
11 6 3 1
35 5 3 1
34 7 3 1
12 1 3 1
36 6 3 1
14 4 3 1
13 4 3 1
37 7 3 1
25 12 2 1
15 1 3 1
39 4 3 1
26 10 2 1
25 1 2 1
3 7 1 1
2 5 3 1
26 6 1 1
28 5 3 1
31 12 2 1
40 13 3 1
5 5 2 1
7 6 3 1
31 5 3 1
17 2 1 1
36 12 2 1
20 9 3 1
19 3 3 1
21 12 1 1
20 4 3 1
22 3 3 1
8 2 3 1
43 13 2 1
9 3 3 1
36 13 3 1
12 11 1 1
24 11 2 1
14 2 3 1
6 9 3 1
26 1 3 1
1 12 3 1
3 1 2 1
2 11 3 1
1 6 3 1
30 11 1 1
29 1 2 1
31 13 2 1
31 2 3 1
41 6 2 1
38 8 2 1
43 7 3 1
42 5 3 1
20 3 3 1
21 2 3 1
32 6 1 1
9 4 1 1
33 7 3 1
23 12 2 1
35 4 2 1
34 4 1 1
12 2 3 1
36 7 1 1
14 5 3 1
13 3 3 1
37 6 3 1
24 5 3 1
39 7 2 1
1 5 3 1
43 6 3 1
3 6 2 1
2 2 3 1
26 9 3 1
4 1 1 1
28 6 3 1
6 4 3 1
5 4 3 1
29 7 1 1
40 7 1 1
16 4 3 1
31 4 2 1
17 1 1 1
19 2 3 1
18 6 3 1
21 11 1 1
20 5 3 1
9 13 2 1
8 3 2 1
23 5 3 1
9 2 3 1
33 13 3 1
11 3 3 1
14 12 3 1
12 4 3 1
14 3 3 1
1 3 3 1
41 3 3 1
6 2 3 1
29 13 3 1
31 3 3 1
41 5 3 1
19 11 1 1
18 1 3 1
42 2 3 1
27 1 2 1
20 12 2 1
43 11 1 1
21 1 2 1
32 7 3 1
9 11 1 1
33 6 3 1
41 7 2 1
30 1 2 1
35 7 3 1
34 5 2 1
33 11 3 1
12 3 3 1
13 2 3 1
37 5 2 1
38 4 3 1
15 3 3 1
39 6 3 1
1 4 3 1
25 7 1 1
2 3 3 1
26 4 2 1
4 2 1 1
28 7 3 1
6 5 3 1
30 6 3 1
29 6 1 1
16 5 1 1
6 8 3 1
36 1 2 1
23 3 3 1
42 13 1 1
16 8 3 1
20 11 1 1
19 5 3 1
18 7 3 1
20 6 3 1
22 1 2 1
21 7 2 1
9 12 2 1
8 4 1 1
23 4 2 1
35 12 2 1
9 1 3 1
33 12 3 1
11 2 3 1
35 1 2 1
34 11 1 1
14 13 2 1
12 5 3 1
13 8 2 1
30 4 3 1
38 7 2 1
25 8 3 1
15 5 3 1
1 2 3 1
27 11 1 1
3 3 2 1
4 4 1 1
26 5 3 1
6 3 3 1
29 12 1 1
7 2 3 1
41 4 3 1
24 6 2 1
40 4 2 1
42 3 3 1
40 6 3 1
16 6 3 1
20 1 2 1
8 11 1 1
7 1 2 1
33 5 3 1
32 10 2 1
35 6 3 1
34 2 2 1
17 6 3 1
11 12 3 1
14 11 2 1
13 1 2 1
37 4 3 1
15 2 3 1
39 1 3 1
38 11 3 1
5 2 3 1
38 6 3 1
25 6 3 1
27 7 1 1
1 8 3 1
36 11 2 1
4 3 3 1
27 4 1 1
30 7 2 1
29 5 1 1
5 7 1 1
5 9 3 1
33 9 3 1
33 7 0 1
22 1 1 2
22 10 3 2
22 7 3 2
22 12 3 2
22 13 3 2
22 8 3 2
22 19 1 2
11 10 3 2
11 13 3 2
11 18 3 2
11 9 3 2
11 19 1 2
11 21 1 2
17 1 3 2
17 2 3 2
17 4 3 2
17 10 3 2
17 9 3 2
17 12 1 2
17 13 3 2
13 1 2 2
13 5 1 2
13 10 3 2
13 12 3 2
13 13 3 2
13 8 3 2
13 9 3 2
9 14 3 2
9 10 3 2
9 9 1 2
9 12 3 2
9 13 3 2
9 16 3 2
9 19 1 2
9 8 3 2
21 1 3 2
21 10 1 2
21 7 3 2
21 13 1 2
21 18 3 2
21 8 3 2
21 9 3 2
27 1 1 2
27 2 1 2
27 3 1 2
27 14 1 2
27 4 1 2
27 10 1 2
27 8 1 2
1 13 3 2
1 9 3 2
1 19 1 2
1 18 3 2
1 21 3 2
1 16 2 2
43 1 2 2
43 10 3 2
43 17 3 2
43 13 3 2
43 12 3 2
43 8 3 2
43 9 3 2
25 1 2 2
25 2 3 2
25 14 1 2
25 4 1 2
25 10 3 2
25 9 3 2
25 12 3 2
25 13 3 2
25 19 1 2
14 11 1 2
14 6 2 2
14 7 3 2
14 13 1 2
14 8 3 2
14 9 3 2
14 19 1 2
24 4 2 2
24 10 3 2
24 6 3 2
24 9 2 2
24 12 3 2
24 13 3 2
24 8 3 2
38 7 2 2
38 12 3 2
38 13 3 2
38 8 3 2
38 9 3 2
38 19 1 2
38 18 1 2
20 1 1 2
20 10 1 2
20 12 3 2
20 13 3 2
20 8 1 2
20 19 1 2
20 18 3 2
6 1 2 2
6 10 3 2
6 12 3 2
6 19 1 2
6 18 3 2
6 21 3 2
34 12 3 2
34 13 3 2
33 16 3 2
2 13 3 2
2 8 3 2
2 9 1 2
2 19 1 2
2 18 2 2
2 21 3 2
3 1 3 2
3 3 3 2
3 4 1 2
3 14 1 2
3 10 1 2
3 6 3 2
3 9 3 2
5 3 3 2
5 5 3 2
5 10 3 2
5 12 3 2
5 13 3 2
5 8 3 2
12 10 1 2
12 12 3 2
12 13 1 2
12 8 3 2
12 9 1 2
12 19 1 2
12 18 3 2
15 10 2 2
15 12 3 2
15 8 3 2
15 9 3 2
15 19 1 2
15 18 3 2
15 21 3 2
18 2 3 2
18 5 3 2
18 10 3 2
18 12 3 2
18 13 3 2
18 8 3 2
18 16 1 2
28 4 3 2
28 10 3 2
28 12 3 2
28 13 3 2
28 8 3 2
29 1 3 2
29 2 1 2
29 3 3 2
29 14 2 2
29 4 3 2
29 10 3 2
29 8 2 2
30 1 3 2
30 10 3 2
30 7 3 2
30 12 3 2
30 13 2 2
30 8 2 2
30 9 3 2
33 8 3 2
33 7 1 2
33 19 1 2
33 18 3 2
33 21 1 2
34 1 3 2
34 2 2 2
34 3 3 2
34 5 1 2
34 14 1 2
34 10 3 2
34 6 3 2
35 1 3 2
35 2 2 2
35 4 3 2
35 10 3 2
35 12 3 2
35 13 3 2
39 10 2 2
39 7 1 2
39 12 3 2
39 13 1 2
39 8 1 2
39 9 3 2
41 1 3 2
41 6 3 2
41 7 3 2
41 13 3 2
41 18 3 2
41 8 2 2
41 9 3 2
29 4 3 2
4 14 2 2
26 9 3 2
37 9 3 2
4 1 2 2
4 2 3 2
4 14 2 2
4 4 1 2
4 8 1 2
4 9 1 2
4 18 2 2
8 3 2 2
8 14 2 2
8 10 2 2
8 9 1 2
8 18 3 2
8 12 3 2
8 8 3 2
19 4 2 2
19 10 3 2
19 8 3 2
19 7 3 2
19 18 2 2
19 12 3 2
19 17 2 2
23 1 3 2
23 14 3 2
23 7 3 2
23 12 3 2
23 18 3 2
23 8 3 2
23 9 1 2
26 4 3 2
26 10 3 2
26 8 1 2
26 18 3 2
26 17 3 2
26 7 1 2
26 19 1 2
31 4 2 2
31 12 3 2
31 13 3 2
31 8 1 2
31 9 1 2
31 19 1 2
31 18 3 2
32 10 3 2
32 8 2 2
32 18 2 2
32 17 1 2

static/js/edi-table.js
/* https://github.com/mindmup/editable-table  */

/*global $, window*/
$.fn.editableTableWidget = function (options) {
	'use strict';
	return $(this).each(function () {
		var buildDefaultOptions = function () {
				var opts = $.extend({}, $.fn.editableTableWidget.defaultOptions);
				opts.editor = opts.editor.clone();
				return opts;
			},
			activeOptions = $.extend(buildDefaultOptions(), options),
			ARROW_LEFT = 37, ARROW_UP = 38, ARROW_RIGHT = 39, ARROW_DOWN = 40, ENTER = 13, ESC = 27, TAB = 9,
			element = $(this),
			editor = activeOptions.editor.css('position', 'absolute').hide().appendTo(element.parent()),
			active,
			showEditor = function (select) {
				active = element.find('td:focus');
				if (active.length && !active.hasClass('static-cell')) {
					editor.val(active.text())
						.removeClass('error')
						.show()
						.offset(active.offset())
						.css(active.css(activeOptions.cloneProperties))
						.width(active.width())
						.height(active.height())
						.focus();
					if (select) {
						editor.select();
					}
				}
			},
			setActiveText = function () {
				var text = editor.val(),
					evt = $.Event('change'),
					originalContent;
				switch(text){
				    case 'z': text = '∅'; break;
				    case 'x': text = '□'; break;
				    case 'c': text = '◩'; break;
				    case 'v': text = '■'; break;
				}
				if (active.text() === text || editor.hasClass('error')) {
					return true;
				}
				originalContent = active.html();
				active.text(text).trigger(evt, text);

				if (evt.result === false) {
					active.html(originalContent);
				}
			},
			movement = function (element, keycode) {
				if (keycode === ARROW_RIGHT) {
					return element.next('td');
				} else if (keycode === ARROW_LEFT) {
					return element.prev('td');
				} else if (keycode === ARROW_UP) {
					return element.parent().prev().children().eq(element.index());
				} else if (keycode === ARROW_DOWN) {
					return element.parent().next().children().eq(element.index());
				}
				return [];
			};
		editor.blur(function () {
			setActiveText();
			editor.hide();
		}).keydown(function (e) {
			if (e.which === ENTER) {
				setActiveText();
				editor.hide();
				active.focus();
				e.preventDefault();
				e.stopPropagation();
			} else if (e.which === ESC) {
				editor.val(active.text());
				e.preventDefault();
				e.stopPropagation();
				editor.hide();
				active.focus();
			} else if (e.which === TAB) {
				active.focus();
			} else if (this.selectionEnd - this.selectionStart === this.value.length || ['z','x','c','v'].indexOf(this.value) >= 0) {
				var possibleMove = movement(active, e.which);
				if (possibleMove.length > 0) {
					possibleMove.focus();
					e.preventDefault();
					e.stopPropagation();
				}
			}
		})
		.on('input paste', function () {
			var evt = $.Event('validate');
			active.trigger(evt, editor.val());
			if (evt.result === false) {
				editor.addClass('error');
			} else {
				editor.removeClass('error');
			}
		});
		element.on('click keypress dblclick', showEditor)
		.css('cursor', 'pointer')
		.keydown(function (e) {
			var prevent = true,
				possibleMove = movement($(e.target), e.which);
			if (possibleMove.length > 0) {
				possibleMove.focus();
			} else if (e.which === ENTER) {
				showEditor(false);
			} else if (e.which === 17 || e.which === 91 || e.which === 93) {
				showEditor(true);
				prevent = false;
			} else {
				prevent = false;
			}
			if (prevent) {
				e.stopPropagation();
				e.preventDefault();
			}
		});

		element.find('td').prop('tabindex', 1);

		$(window).on('resize', function () {
			if (editor.is(':visible')) {
				editor.offset(active.offset())
				.width(active.width())
				.height(active.height());
			}
		});
	});

};
$.fn.editableTableWidget.defaultOptions = {
	cloneProperties: ['padding', 'padding-top', 'padding-bottom', 'padding-left', 'padding-right',
					  'text-align', 'font', 'font-size', 'font-family', 'font-weight',
					  'border', 'border-top', 'border-bottom', 'border-left', 'border-right'],
	editor: $('<input>')
};

